### ISSUE TITLE
Starting upload when the download finished

### ISSUE THREAD

**Issue by unabomber87** [akinaru/speed-test-lib#9]
Hi, i would first thank you for this useful library.
Here's my problem i want to trigger the upload process when the download is finished.
So on the `ISpeedTestListener` `onDownloadPacketsReceived` method i have putted the `speedTestSocket.startUpload("XXXXXX.com", 80, "/upload/upload.php", 10000000)` but it doesn't start.
Is it a bug? or am i doing something wrong?

Thanks

**Comment by akinaru** [comment]
Hi, It should work, for instance the following snippet makes an upload following a download on `2.testdebit.info` : 

```
public class DownloadFileExample {

    /**
     * socket timeout used in ms.
     */
    private final static int SOCKET_TIMEOUT = 5000;

    /**
     * speed examples server host name.
     */
    private final static String SPEED_TEST_SERVER_HOST = "2.testdebit.info";

    /**
     * spedd examples server uri.
     */
    private final static String SPEED_TEST_SERVER_URI_DL = "/fichiers/1Mo.dat";

    /**
     * speed examples server port.
     */
    private final static int SPEED_TEST_SERVER_PORT = 80;

    /**
     * upload 10Mo file size.
     */
    private static final int FILE_SIZE = 1000000;

    /**
     * spedd examples server uri.
     */
    private static final String SPEED_TEST_SERVER_URI_UL = "/";
    
    /**
     * Download file example main.
     *
     * @param args no args required
     */
    public static void main(final String[] args) {

        // instantiate speed examples
        final SpeedTestSocket speedTestSocket = new SpeedTestSocket();

        //set timeout for download
        speedTestSocket.setSocketTimeout(SOCKET_TIMEOUT);

        // add a listener to wait for speed examples completion and progress
        speedTestSocket.addSpeedTestListener(new ISpeedTestListener() {

            @Override
            public void onDownloadPacketsReceived(final long packetSize, final BigDecimal transferRateBitPerSeconds,
                                                  final BigDecimal transferRateOctetPerSeconds) {

                System.out.println("----------onDownloadPacketsReceived----------");
                System.out.println("transferRateBitPerSeconds : " + transferRateBitPerSeconds);

                speedTestSocket.startUpload(SPEED_TEST_SERVER_HOST, SPEED_TEST_SERVER_PORT, SPEED_TEST_SERVER_URI_UL,
                        FILE_SIZE);

            }

            @Override
            public void onDownloadError(final SpeedTestError speedTestError, final String errorMessage) {
                System.out.println("download error : " + speedTestError);
            }

            @Override
            public void onUploadPacketsReceived(final long packetSize, final BigDecimal transferRateBitPerSeconds, final
            BigDecimal transferRateOctetPerSeconds) {

                System.out.println("----------onUploadPacketsReceived----------");
                System.out.println("transferRateBitPerSeconds : " + transferRateBitPerSeconds);

            }

            @Override
            public void onUploadError(final SpeedTestError speedTestError, final String errorMessage) {
                System.out.println("upload error : " + speedTestError);
            }

            @Override
            public void onDownloadProgress(final float percent, final SpeedTestReport downloadReport) {
                System.out.println("----------onDownloadProgress----------");
                System.out.println("transferRateBitPerSeconds : " + downloadReport.getTransferRateBit());
            }

            @Override
            public void onUploadProgress(final float percent, final SpeedTestReport uploadReport) {
                System.out.println("----------onUploadProgress----------");
                System.out.println("transferRateBitPerSeconds : " + uploadReport.getTransferRateBit());
            }
        });

        speedTestSocket.startDownload(SPEED_TEST_SERVER_HOST, SPEED_TEST_SERVER_PORT, SPEED_TEST_SERVER_URI_DL);
    }
}
```
Is your upload working when it is not queued after a download ?

**Comment by unabomber87** [comment]
Hi yes it work when it isn't queued.
I dont have the same signature for the 2 method `onDownloadPacketsReceived` and `onUploadPacketsReceived`
`@Override
public void onDownloadPacketsReceived(long packetSize,
		float transferRateBps, float transferRateOps) {
	// TODO Auto-generated method stub
}

@Override
public void onUploadPacketsReceived(long packetSize,
		float transferRateBps, float transferRateOps) {
	// TODO Auto-generated method stub
}`
is it normal?

**Comment by akinaru** [comment]
Yes sorry this is `float` instead of `BigDecimal`

**Comment by unabomber87** [comment]
here's the code that i use 
`
// instantiate speed examples
final SpeedTestSocket speedTestSocket = new SpeedTestSocket();

//set timeout for download
speedTestSocket.setSocketTimeout(SOCKET_TIMEOUT);

// add a listener to wait for speed examples completion and progress
speedTestSocket.addSpeedTestListener(new ISpeedTestListener() {


@Override
public void onDownloadError(final SpeedTestError speedTestError, final String errorMessage) {
    System.out.println("download error : " + speedTestError);
}



@Override
public void onUploadError(final SpeedTestError speedTestError, final String errorMessage) {
    System.out.println("upload error : " + speedTestError);
}

@Override
public void onDownloadProgress(final float percent, final SpeedTestReport downloadReport) {
//                System.out.println("----------onDownloadProgress----------");
//                System.out.println("transferRateBitPerSeconds : " + downloadReport.getTransferRateBit());
}

@Override
public void onUploadProgress(final float percent, final SpeedTestReport uploadReport) {
//                System.out.println("----------onUploadProgress----------");
//                System.out.println("transferRateBitPerSeconds : " + uploadReport.getTransferRateBit());
}

@Override
public void onDownloadPacketsReceived(long packetSize,float transferRateBps, float transferRateOps) {
	// TODO Auto-generated method stub
	System.out.println("transferRateBitPerSeconds : " + transferRateBps);
	speedTestSocket.startUpload(SPEED_TEST_SERVER_HOST, SPEED_TEST_SERVER_PORT, SPEED_TEST_SERVER_URI_UL,FILE_SIZE);

}

@Override
public void onUploadPacketsReceived(long packetSize,float transferRateBps, float transferRateOps) {
	// TODO Auto-generated method stub
	System.out.println("transferRateBitPerSeconds : " + transferRateBps);

}
});

speedTestSocket.startDownload(SPEED_TEST_SERVER_HOST, SPEED_TEST_SERVER_PORT, SPEED_TEST_SERVER_URI_DL);
`
the problem here is that the method `onUploadPacketsReceived` doesn't trigger when upload finish
but the method `onUploadProgress` show the progress during the upload

**Comment by akinaru** [comment]
I see, that is not the expected behavior. Will get into it thanks. To workaround this run upload in a new thread as : 

```
new Thread(new Runnable() {
    @Override
    public void run() {
        speedTestSocket.startUpload(SPEED_TEST_SERVER_HOST, SPEED_TEST_SERVER_PORT, SPEED_TEST_SERVER_URI_UL,
                FILE_SIZE);
    }
}).start();
```

**Comment by unabomber87** [comment]
thanks a lot for you help :)

**Comment by akinaru** [comment]
This bug is now fixed in v`1.21`
